dynamiq/memory/backend/sqlite.py

```diff
@@ -0,0 +1,196 @@
+
+
+class SQLiteError(Exception):
+    """Base exception class for SQLite-related errors in the memory backend."""
+    pass
+
+
+class SQLite(MemoryBackend):
+    """SQLite implementation of the memory storage backend."""
+
+    name = "SQLite"
+
+    # Define constants for the SQL queries
+    CREATE_TABLE_QUERY = """
+        CREATE TABLE IF NOT EXISTS {index_name} (
+            id TEXT PRIMARY KEY,
+            role TEXT NOT NULL,
+            content TEXT NOT NULL,
+            metadata TEXT,
+            timestamp REAL
+        )
+    """  # nosec B608
+
+    VALIDATE_TABLE_QUERY = "SELECT name FROM sqlite_master WHERE type='table' AND name=?"
+    INSERT_MESSAGE_QUERY = """
+        INSERT INTO {index_name} (id, role, content, metadata, timestamp)
+        VALUES (?, ?, ?, ?, ?)
+    """  # nosec B608
+    SELECT_ALL_MESSAGES_QUERY = """
+        SELECT id, role, content, metadata, timestamp
+        FROM {index_name}
+        ORDER BY timestamp ASC
+    """  # nosec B608
+    CHECK_IF_EMPTY_QUERY = "SELECT COUNT(*) FROM {index_name}"  # nosec B608
+    CLEAR_TABLE_QUERY = "DELETE FROM {index_name}"  # nosec B608
+    SEARCH_MESSAGES_QUERY = """
+        SELECT id, role, content, metadata
+        FROM {index_name}
+    """  # nosec B608
+
+    def __init__(self, db_path: str = "conversations.db", index_name: str = "conversations"):
+        """Initializes the SQLite memory storage."""
+        self.db_path = db_path
+        self.index_name = index_name
+
+        try:
+            self._validate_table_name(create_if_not_exists=True)
+        except Exception as e:
+            raise SQLiteError(f"Error initializing SQLite backend: {e}") from e
+
+    def _validate_table_name(self, create_if_not_exists: bool = False):
+        """Validates the table name to prevent SQL injection and optionally creates it."""
+        if not re.match(r"^[A-Za-z0-9_]+$", self.index_name):
+            raise SQLiteError(f"Invalid table name: '{self.index_name}'")
+
+        try:
+            with sqlite3.connect(self.db_path) as conn:
+                cursor = conn.cursor()
+                cursor.execute(self.VALIDATE_TABLE_QUERY, (self.index_name,))  # nosec B608
+                result = cursor.fetchone()
+
+                if result is None:
+                    if create_if_not_exists:
+                        self._create_table()
+                    else:
+                        raise SQLiteError(f"Table '{self.index_name}' does not exist in the database.")
+
+        except sqlite3.Error as e:
+            raise SQLiteError(f"Error validating or creating table: {e}") from e
+
+    def _create_table(self):
+        """Creates the messages table."""
+        query = self.CREATE_TABLE_QUERY.format(index_name=self.index_name)
+        try:
+            with sqlite3.connect(self.db_path) as conn:
+                cursor = conn.cursor()
+                cursor.execute(query)  # nosec B608
+                conn.commit()
```

Comment:
I recommend creating a separate method for the execute function and reusing it everywhere. I have found my old SQLite helper; maybe we can reuse it
```
import sqlite3
from sqlite3 import Cursor
from typing import Dict, List, Tuple


class SqliteClient:
    _connections = {}

    def __init__(self, db_path: str):
        """
        Initializes a new instance of the SqliteClient class and creates
        a connection to the SQLite database, if it not exists yet

        Parameters:
        db_path (str): The path of the SQLite database.
        """
        self.db_path = db_path
        self._init_connection(db_path)

    def __del__(self):
        """Closes the SQLite connection when the object is destroyed"""
        if self._connections.get(self.db_path) is not None:
            self._connections[self.db_path].close()
            self._connections.pop(self.db_path)

    def _init_connection(self, db_path: str):
        if self._connections.get(db_path) is None:
            self._connections[db_path] = sqlite3.connect(
                db_path, check_same_thread=False, timeout=10
            )
        self.cursor = self._connections[db_path].cursor()

    def execute(self, query: str, params: Tuple = (), commit: bool = False) -> Cursor:
        self._init_connection(self.db_path)
        executed_cursor = self.cursor.execute(query, params)

        if commit:
            self._connections[self.db_path].commit()

        return executed_cursor

    def query_all(self, query: str, params: Tuple = ()) -> List[Dict]:
        """
        Executes a query on the SQLite database and returns all the results.

        Parameters:
        query: The SQL query to be executed.
        params: query parameters

        Returns:
        list: A list of dicts containing the results of the query
        """
        executed_cursor = self.execute(query, params)
        columns = [column[0] for column in executed_cursor.description]
        query_res = [dict(zip(columns, row)) for row in executed_cursor.fetchall() or []]
        return query_res

    def query_one(self, query: str, params: Tuple = ()) -> Dict:
        """
        Executes a query on the SQLite database and returns the first result.

        Parameters:
        query: The SQL query to be executed.
        params: query parameters

        Returns:
        tuple: A dict containing the first result of the query
        """
        executed_cursor = self.execute(query, params)
        columns = [column[0] for column in executed_cursor.description]
        query_res = dict(zip(columns, executed_cursor.fetchone() or []))
        return query_res

    def table_exists(self, table_name):
        query = "SELECT name FROM sqlite_master WHERE type='table' AND name=?"
        return self.query_one(query, (table_name,))
```
And example of use
```
self.sqlite = SqliteClient(db_path=db_path)
self.sqlite.query_all(SQL_QUERY, (self.index_name,))
```